test/engine-luatest/gh_4973_concurrent_alter_fails_test.lua

```diff
@@ -0,0 +1,53 @@
+local server = require('luatest.server')
+local t = require('luatest')
+
+local g = t.group('Replace during primary key altering')
+
+g.before_all(function()
+    g.server = server:new({})
+    g.server:start()
+end)
+
+g.after_all(function()
+    g.server:stop()
+    g.server = nil
+end)
+
+g.test_concurrent_alter = function()
+    g.server:exec(function()
+        local fiber = require('fiber')
+        local s = box.schema.space.create('test', {engine = 'memtx'})
```

Comment:
The reason is `vinyl` does not support rebuilding the primary index of a non-empty space.

Reply:
OK, I see. Still, let's make this test multi-engine and use `t.skip_if` with a valid reason to skip the vinyl test so that it's obvious why the test is memtx-only.

Reply:
Done.

<details><summary>diff</summary>

```diff
diff --git a/test/engine-luatest/gh_4973_concurrent_alter_fails_test.lua b/test/engine-luatest/gh_4973_concurrent_alter_fails_test.lua
index 725f7fa502d0..28f49c4e999c 100644
--- a/test/engine-luatest/gh_4973_concurrent_alter_fails_test.lua
+++ b/test/engine-luatest/gh_4973_concurrent_alter_fails_test.lua
@@ -1,27 +1,32 @@
 local server = require('luatest.server')
 local t = require('luatest')
 
-local g = t.group()
+local g = t.group(nil, {
+    {engine = 'memtx'},
+    {engine = 'vinyl'},
+})
 
-g.before_all(function()
-    g.server = server:new({})
-    g.server:start()
+g.before_all(function(cg)
+    t.skip_if(cg.params.engine == 'vinyl',
+              "Vinyl doesn't support rebuilding primary key")
+    cg.server = server:new({})
+    cg.server:start()
 end)
 
-g.after_all(function()
-    g.server:stop()
-    g.server = nil
+g.after_all(function(cg)
+    cg.server:stop()
+    cg.server = nil
 end)
 
-g.before_each(function()
-    g.server:exec(function()
-        local s = box.schema.space.create('test', {engine = 'memtx'})
+g.before_each(function(cg)
+    cg.server:exec(function(engine)
+        local s = box.schema.space.create('test', {engine = engine})
         s:create_index('pk')
-    end)
+    end, {cg.params.engine})
 end)
 
-g.after_each(function()
-    g.server:exec(function()
+g.after_each(function(cg)
+    cg.server:exec(function()
         -- If some tuples are not referenced by primary index, they will be
         -- deleted by gc and we will catch segmentation fault on s:drop().
         collectgarbage("collect")
@@ -29,8 +34,8 @@ g.after_each(function()
     end)
 end)
 
-g.test_concurrent_alter = function()
-    g.server:exec(function()
+g.test_concurrent_alter = function(cg)
+    cg.server:exec(function()
         local fiber = require('fiber')
         local tuple_num = 10000
         local iteration_steps = 10
@@ -51,18 +56,18 @@ g.test_concurrent_alter = function()
         end
 
         local alter_is_over = false
-        local function disturb()
-            while not alter_is_over do
-                box.atomic(do_random_updates)
-                fiber.sleep(0)
-            end
-        end
-
         local function alter()
             s.index.pk:alter({parts = {{field = 2, type = 'unsigned'}}})
             alter_is_over = true
         end
 
+        local function disturb()
+            while not alter_is_over do
+                box.atomic(do_random_updates)
+                fiber.sleep(0)
+            end
+        end
+
         local function collect_garbage()
             while not alter_is_over do
                 fiber.sleep(0)
@@ -78,8 +83,8 @@ g.test_concurrent_alter = function()
         disturber:set_joinable(true)
         alterer:set_joinable(true)
 
-        t.assert(gc:join())
-        t.assert(disturber:join())
         t.assert(alterer:join())
+        t.assert(disturber:join())
+        t.assert(gc:join())
     end)
 end
```

</details>

Reply:
Done.

<details><summary>diff</summary>

```diff
diff --git a/test/engine-luatest/gh_4973_concurrent_alter_fails_test.lua b/test/engine-luatest/gh_4973_concurrent_alter_fails_test.lua
index 725f7fa502d0..28f49c4e999c 100644
--- a/test/engine-luatest/gh_4973_concurrent_alter_fails_test.lua
+++ b/test/engine-luatest/gh_4973_concurrent_alter_fails_test.lua
@@ -1,27 +1,32 @@
 local server = require('luatest.server')
 local t = require('luatest')
 
-local g = t.group()
+local g = t.group(nil, {
+    {engine = 'memtx'},
+    {engine = 'vinyl'},
+})
 
-g.before_all(function()
-    g.server = server:new({})
-    g.server:start()
+g.before_all(function(cg)
+    t.skip_if(cg.params.engine == 'vinyl',
+              "Vinyl doesn't support rebuilding primary key")
+    cg.server = server:new({})
+    cg.server:start()
 end)
 
-g.after_all(function()
-    g.server:stop()
-    g.server = nil
+g.after_all(function(cg)
+    cg.server:stop()
+    cg.server = nil
 end)
 
-g.before_each(function()
-    g.server:exec(function()
-        local s = box.schema.space.create('test', {engine = 'memtx'})
+g.before_each(function(cg)
+    cg.server:exec(function(engine)
+        local s = box.schema.space.create('test', {engine = engine})
         s:create_index('pk')
-    end)
+    end, {cg.params.engine})
 end)
 
-g.after_each(function()
-    g.server:exec(function()
+g.after_each(function(cg)
+    cg.server:exec(function()
         -- If some tuples are not referenced by primary index, they will be
         -- deleted by gc and we will catch segmentation fault on s:drop().
         collectgarbage("collect")
@@ -29,8 +34,8 @@ g.after_each(function()
     end)
 end)
 
-g.test_concurrent_alter = function()
-    g.server:exec(function()
+g.test_concurrent_alter = function(cg)
+    cg.server:exec(function()
         local fiber = require('fiber')
         local tuple_num = 10000
         local iteration_steps = 10
@@ -51,18 +56,18 @@ g.test_concurrent_alter = function()
         end
 
         local alter_is_over = false
-        local function disturb()
-            while not alter_is_over do
-                box.atomic(do_random_updates)
-                fiber.sleep(0)
-            end
-        end
-
         local function alter()
             s.index.pk:alter({parts = {{field = 2, type = 'unsigned'}}})
             alter_is_over = true
         end
 
+        local function disturb()
+            while not alter_is_over do
+                box.atomic(do_random_updates)
+                fiber.sleep(0)
+            end
+        end
+
         local function collect_garbage()
             while not alter_is_over do
                 fiber.sleep(0)
@@ -78,8 +83,8 @@ g.test_concurrent_alter = function()
         disturber:set_joinable(true)
         alterer:set_joinable(true)
 
-        t.assert(gc:join())
-        t.assert(disturber:join())
         t.assert(alterer:join())
+        t.assert(disturber:join())
+        t.assert(gc:join())
     end)
 end
```

</details>